### ISSUE TITLE
[Feature Request]: Add ability to build multiple projects on the command line without sln or slnf

### ISSUE THREAD

**Issue by LostTime76** [dotnet/msbuild#9714]
### Summary

I would like to know the background story of why this is not possible to do today and figured I would ask if it was possible to make this happen. 

### Background and Motivation

Solution files...................

Today I can build multiple projects with dotnet build by passing in a solution or solution filter to specify the projects I want to build. Additionally, I have the ability to build a single project without a sln or slnf by passing in the name of the project directly. Therefore,  it seems like there is no limitation on building multiple projects in a single dotnet build call. 

The only reason I have had to create solution files is so that I can have the ability to build multiple related projects at a time. That's it, they do nothing more and they clutter the file tree. Recently, I have had to be able to 'select' individual projects or groups of projects to build within a file tree. This has led me to solution filters, which feel like a cludge over already crusty solution files.

Furthermore, I can invoke the MSBuild task within a project file and pass to it an item list of projects to build! This further solidifies my belief that there is no limitation of building without solution files.

### Proposed Feature

Add the ability to pass a list of projects to the dotnet cli / msbuild cli to build without a solution file instead of being limited by a single project input. I think this would get rid of the requirement for many people to need solution files. It can make build automation and selecting the projects to build much easier by allowing them just to be specified on the command line directly.

Proposal(s):

1. Allow the user to specify multiple project paths on the command line in the invocation. This would just be an extended feature of the existing command line.
2. If (1) is not possible, provide a new command line option for specifying multiple projects. The 'how' is not important. Could be a single option delimited by a special character or a repeated option such as: "-proj proj1path.proj -proj proj2path.proj"

**Comment by naeemaei** [comment]
Can I work on this feature?

**Comment by baronfel** [comment]
This seems like a logical and natural extension of the existing MSBuild CLI Grammar - we'd love to see a pull request in this space!

There are a few things to think about here:
* There are multiple ways to specify a project - either explicitly via passing project files as command line arguments or implicitly discovering a project from the current working directory. We should define how the build should behave when 'implicit discovery' occurs
  * **Proposal**: any change to allow multiple projects should only apply when multiple projects are provided directly on the command line - there should be _no change_ to implicit discovery.
* MSbuild can accept .binlog files as an input - this allows for 'replaying' a previous build, often with different loggers/analysis enabled. We should define how the build would run if both binlog inputs and project files were specified.
  * **Proposal**: if both .binlog and project files are specified at the command line, an error should be thrown instructing the user to provide only a single binlog file or one or more project files.
* MSbuild can accept solution files, solution filters, and project files on the command line. We should define what the expected behavior is when combinations of these types of files are specified.
  * **Proposal**: Only multiple project files should be allowed to be passed - multiple solutions or solution filters, and combinations of solutions/solution filters/project files should raise an error saying that only multiple project files are allowed to be passed to MSbuild.
    * **Reasoning**: solutions and solution filters are converted to a 'metaproject' with a lot of implicit behaviors that don't directly align to project-file-based usage of MSBuild - it is unclear how to reasonable combine/merge these different kinds of inputs.

Useful links
* [here is where the project argument is defined](https://github.com/dotnet/msbuild/blob/5768954b7fac850d5a721ea4c4e37403564059c2/src/MSBuild/CommandLineSwitches.cs#L235) - this will likely need to be changed to remove the `DuplicateSwitchError` and allow `multi params`

**Comment by baronfel** [comment]
@naeemaei I realized I forgot to @ you directly - if you'd like to contribute this feature we'd love to help you implement it!

**Comment by naeemaei** [comment]
@baronfel Yes, I'd like to contribute.

**Comment by zivkan** [comment]
This issue has label gathering-feedback, so if there's any specific feedback you'd like, please let me know. In short, I feel being able to pass multiple projects to a single CLI invocation and have it automatically do a P2P traversal (something that I think does not happen with slnf files) would help simplify CI scripts for my complex and multi-pipeline product.

**Comment by baronfel** [comment]
No feedback required at this point, just need to do the work. 

I have talked to other teams like DevKit and it would help their story for not needing to generate a solution file to enable builds in repos that don't have them.

**Comment by baronfel** [comment]
The big blocker here that will likely limit how useful this is in practice is that any properties specified on the command line will apply as global properties to _all_ specified projects to be built - this means common operations like specifying a targetframework or runtime identifier will likely not work as folks expect, since all projects would need to support that TFM or that RID.  This could still be useful for common workflows like "just build all these things"

**Comment by Frulfump** [comment]
> The big blocker here that will likely limit how useful this is in practice is that any properties specified on the command line will apply as global properties to _all_ specified projects to be built - this means common operations like specifying a targetframework or runtime identifier will likely not work as folks expect, since all projects would need to support that TFM or that RID.  This could still be useful for common workflows like "just build all these things"

Good point. Not an issue for my use cases so hope it can still be provided even in that "limited" form.

**Comment by zivkan** [comment]
> The big blocker here that will likely limit how useful this is in practice is that any properties specified on the command line will apply as global properties to _all_ specified projects to be built

I don't understand how it's a blocker.  It's how using a solution file, or solution filter works, no?

**Comment by baronfel** [comment]
@zivkan not exactly - the generated metaproj for a solution has a lot of additional metadata controlling which properties can be set/unset/treated as local properties that helps paper over the global properties you pass in.

**Comment by acnicholls-kroll** [comment]
I am just now running into this.  we have globbing patterns in pipeline tasks that use for `DotNetCoreCLI@2` and trying to replicate issues with the running commands, because the DevOps action allows globs, but CLI doesn't, means that it's not easy to repro issues and then find fixes.

**Comment by Frulfump** [comment]
> I am just now running into this. we have globbing patterns in pipeline tasks that use for `DotNetCoreCLI@2` and trying to replicate issues with the running commands, because the DevOps action allows globs, but CLI doesn't, means that it's not easy to repro issues and then find fixes.

I've had that exact issue in the past and is why I have stopped using `DotNetCoreCLI@2` in my Azure DevOps Azure Pipelines and only use plain `dotnet` commands and don't want the extra abstraction from a task where it's unclear how it actually works (One issue is that build warnings aren't surfaced as you would expect, so I follow what's outlined in this repo for that https://github.com/melix-dev/azure-devops-dotnet-warnings). It's so worth it to be able to just copy the command that ran from CI into a new clone of the repo that's having issues to see if you can repro the issues. (Having a binlog created in CI helps a lot too).

I only want to publish a top-level project anyways and `dotnet build` I can run against the sln (or could run two build commands one for app project and one for test project, or just the test project if you test your top-level project in that as it would be built then but I like having it separated to make it easier to find restore/build issues with separate restore and build commands for both projects) we do have but would be good with glob support for `dotnet build` or this issue.
